tests/gtest/avifincrtest_helpers.cc

```diff
@@ -0,0 +1,303 @@
+        --gridCols;
+        *cellWidth = image.width / gridCols;
+    }
+    while ((gridRows > 1) &&
+           (((*cellHeight * gridRows) != image.height) || (*cellHeight < 64) || (needEvenHeights && ((*cellHeight & 1) != 0)))) {
+        --gridRows;
+        *cellHeight = image.height / gridRows;
+    }
+
+    std::vector<testutil::avifImagePtr> cellImages;
+    cellImages.reserve(gridCols * gridRows);
+    for (uint32_t row = 0, iCell = 0; row < gridRows; ++row) {
+        for (uint32_t col = 0; col < gridCols; ++col, ++iCell) {
+            avifCropRect cell;
+            cell.x = col * *cellWidth;
+            cell.y = row * *cellHeight;
+            cell.width = ((cell.x + *cellWidth) <= image.width) ? *cellWidth : (image.width - cell.x);
+            cell.height = ((cell.y + *cellHeight) <= image.height) ? *cellHeight : (image.height - cell.y);
+            cellImages.emplace_back(avifImageCreateEmpty(), avifImageDestroy);
+            ASSERT_NE(cellImages.back(), nullptr);
+            ASSERT_EQ(avifImageSetViewRect(cellImages.back().get(), &image, &cell), AVIF_RESULT_OK);
+        }
+    }
+
+    testutil::avifEncoderPtr encoder(avifEncoderCreate(), avifEncoderDestroy);
+    ASSERT_NE(encoder, nullptr);
+    encoder->speed = AVIF_SPEED_FASTEST;
+    std::vector<avifImage *> cellImagePtrs(cellImages.size()); // Just here to match libavif API.
+    for (size_t i = 0; i < cellImages.size(); ++i) {
+        cellImagePtrs[i] = cellImages[i].get();
+    }
+    ASSERT_EQ(avifEncoderAddImageGrid(encoder.get(), gridCols, gridRows, cellImagePtrs.data(), AVIF_ADD_IMAGE_FLAG_SINGLE), AVIF_RESULT_OK);
+    ASSERT_EQ(avifEncoderFinish(encoder.get(), output), AVIF_RESULT_OK);
+}
+
+// Encodes the image to be decoded incrementally.
+void encodeAsIncremental(const avifImage & image, bool flatCells, avifRWData * output, uint32_t * cellWidth, uint32_t * cellHeight)
+{
+    const uint32_t gridCols = image.width / 64; // 64px is the min cell width.
+    const uint32_t gridRows = flatCells ? 1 : (image.height / 64);
+    encodeAsGrid(image, (gridCols > 1) ? gridCols : 1, (gridRows > 1) ? gridRows : 1, output, cellWidth, cellHeight);
+}
+
+} // namespace
+
+void encodeRectAsIncremental(const avifImage & image,
+                             uint32_t width,
+                             uint32_t height,
+                             bool createAlphaIfNone,
+                             bool flatCells,
+                             avifRWData * output,
+                             uint32_t * cellWidth,
+                             uint32_t * cellHeight)
+{
+    avifImagePtr subImage(avifImageCreateEmpty(), avifImageDestroy);
+    ASSERT_NE(subImage, nullptr);
+    ASSERT_LE(width, image.width);
+    ASSERT_LE(height, image.height);
+    avifPixelFormatInfo info;
+    avifGetPixelFormatInfo(image.yuvFormat, &info);
+    const avifCropRect rect { .x = ((image.width - width) / 2) & ~info.chromaShiftX,
```

Comment:
designated initializers are a c++20 feature

Reply:
Ah yes, having them in C99 confused me.